File: app/src/main/java/libertypassage/com/app/basic/UpdateAddress.java
package libertypassage.com.app.basic;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.gson.Gson;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import libertypassage.com.app.R;
import libertypassage.com.app.models.ModelResponse;
import libertypassage.com.app.other.QRCodeScanerActivity;
import libertypassage.com.app.utilis.ApiInterface;
import libertypassage.com.app.utilis.ClientInstance;
import libertypassage.com.app.utilis.Constants;
import libertypassage.com.app.utilis.Utility;
import libertypassage.com.app.widgets.CenturyGothicEditText;
import libertypassage.com.app.widgets.CenturyGothicTextview;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UpdateAddress extends AppCompatActivity implements View.OnClickListener {

    private final String TAG = UpdateAddress.class.getSimpleName();
    private CenturyGothicTextview tv_next, tv_currentLocation;
    private CenturyGothicEditText et_address;
    private ImageView iv_back, iv_edit, iv_qrcode_scanerr;
    private String token, from, s_address = "", postalCode = "", city = "", tempType, currentTemp;
    private Context context;
    private FusedLocationProviderClient mFusedLocationClient;
    private double latitudes, longitudes, altitudes;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.upadate_address);

        context = UpdateAddress.this;
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
        findID();
    }


    private void findID() {
        iv_back = findViewById(R.id.iv_back);
        iv_qrcode_scanerr = findViewById(R.id.iv_qrcode_scanerr);
        et_address = findViewById(R.id.et_address);
        tv_currentLocation = findViewById(R.id.tv_currentLocation);
        iv_edit = findViewById(R.id.iv_edit);
        tv_next = findViewById(R.id.tv_next);

        iv_back.setOnClickListener(this);
        iv_qrcode_scanerr.setOnClickListener(this);
        tv_currentLocation.setOnClickListener(this);
        iv_edit.setOnClickListener(this);
        tv_next.setOnClickListener(this);
        token = Utility.getSharedPreferences(context, Constants.KEY_BEARER_TOKEN);
        tempType = Utility.getSharedPreferences(context, Constants.KEY_TEMP_TYPE);
        currentTemp = Utility.getSharedPreferences(context, Constants.KEY_MY_TEMP);


        from = getIntent().getStringExtra("from");
        et_address.setText(Utility.getSharedPreferences(context, Constants.KEY_CURRENT_LOC));
    }


    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.iv_back: {
                finish();
                break;
            }
            case R.id.iv_edit: {
                et_address.setEnabled(true);
                et_address.setSelection(et_address.length());
                break;
            }
            case R.id.tv_next: {
                s_address = et_address.getText().toString().trim();
                if (s_address.equals("")) {
                    Toast.makeText(context, "Please add address", Toast.LENGTH_LONG).show();
                } else if (Utility.isConnectingToInternet(context)) {

                    updateAddressAndTemp();

                } else {
                    Toast.makeText(context, "Please connect the internet", Toast.LENGTH_LONG).show();
                }
                break;
            }
            case R.id.tv_currentLocation: {
                dialogUpdate();
                break;
            }
            case R.id.iv_qrcode_scanerr: {
                Intent intent = new Intent(context, QRCodeScanerActivity.class);
                startActivityForResult(intent, 170);
                break;
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 170) {
            if (resultCode == 160) {
                String contents = data.getStringExtra("ResultText");
                if (contents != null) {
                    et_address.setText(contents);
                    et_address.setSelection(et_address.length());
                    Log.e("contents", contents);
                }
            }

            if (resultCode == RESULT_CANCELED) {
                //handle cancel
            }
        }
    }

    private void dialogUpdate() {
        Dialog dialog;
        dialog = new Dialog(UpdateAddress.this);
        dialog.setContentView(R.layout.dialog_get_location);
        dialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.setCancelable(false);

        ImageView cancel = dialog.findViewById(R.id.cancel);
        CenturyGothicTextview tv_notAllow = dialog.findViewById(R.id.tv_notAllow);
        CenturyGothicTextview tv_ok = dialog.findViewById(R.id.tv_ok);

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
            }
        });

        tv_notAllow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
            }
        });

        tv_ok.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getLastLocation();
                dialog.dismiss();
            }
        });

        dialog.show();
    }

    @SuppressLint("MissingPermission")
    private void getLastLocation() {
        LocationRequest locationRequest = LocationRequest.create();
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
        mFusedLocationClient.getLastLocation().addOnCompleteListener(
                new OnCompleteListener<Location>() {
                    @Override
                    public void onComplete(@NonNull Task<Location> task) {
                        Location location = task.getResult();
                        if (location != null) {
                            latitudes = location.getLatitude();
                            longitudes = location.getLongitude();
                            altitudes = location.getAltitude();
                            Log.e("altitudes", altitudes+"");
                            getAddress();
                        }
                    }
                }
        );
    }

    private void getAddress() {
        Geocoder geocoder = new Geocoder(context, Locale.getDefault());
        List<Address> addresses;
        try {
            addresses = geocoder.getFromLocation(latitudes, longitudes, 1); // Here 1 represent max location result to returned, by documents it recommended 1 to 5
            String address11 = addresses.get(0).getAddressLine(0); // If any additional address line present than only, check with max available address lines by getMaxAddressLineIndex()
//            String city11 = addresses.get(0).getLocality();
//            state = addresses.get(0).getAdminArea();
//            country = addresses.get(0).getCountryName();
//            postalCode = addresses.get(0).getPostalCode();
//            knownName = addresses.get(0).getFeatureName(); // Only if available else return NULL
            et_address.setText(address11);
            et_address.setSelection(et_address.length());
        } catch (IOException e) {

            e.printStackTrace();
        }
    }

    private void addUserAddress() {
        Utility.showProgressDialog(context);
        ApiInterface apiInterface = ClientInstance.getRetrofitInstance().create(ApiInterface.class);
        Call<ModelResponse> call = apiInterface.addUserAddress(Constants.KEY_HEADER + token, Constants.KEY_BOT, s_address, postalCode, city);

        call.enqueue(new Callback<ModelResponse>() {
            @Override
            public void onResponse(Call<ModelResponse> call, Response<ModelResponse> response) {
                Utility.stopProgressDialog(context);
                ModelResponse model = response.body();
                Log.e("AddUserAdd", new Gson().toJson(model));
                if (model != null && model.getError().equals(false)) {

                    Toast.makeText(context, model.getMessage(), Toast.LENGTH_LONG).show();
                    Utility.setSharedPreference(context, Constants.KEY_CURRENT_LOC, s_address);

                    Intent intent = getIntent();
                    intent.putExtra("address", s_address);
                    setResult(121, intent);
                    finish();


                } else if (model != null && model.getError().equals(true)) {
                    Utility.stopProgressDialog(context);
                    Toast.makeText(context, model.getMessage(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<ModelResponse> call, Throwable t) {
                Utility.stopProgressDialog(context);
//                Log.e("model", "onFailure    " + t.getMessage());
                Toast.makeText(context, Constants.ERROR_MSG, Toast.LENGTH_LONG).show();
            }
        });
    }

    private void updateAddressAndTemp() {
        Utility.showProgressDialog(context);
        ApiInterface apiInterface = ClientInstance.getRetrofitInstance().create(ApiInterface.class);
        Call<ModelResponse> call = apiInterface.addressAndTemp(Constants.KEY_HEADER + token,
                Constants.KEY_BOT, s_address, tempType, currentTemp);

        call.enqueue(new Callback<ModelResponse>() {
            @Override
            public void onResponse(Call<ModelResponse> call, Response<ModelResponse> response) {
                Utility.stopProgressDialog(context);
                ModelResponse model = response.body();

                if (model != null && model.getError().equals(false)) {
                    Toast.makeText(context, model.getMessage(), Toast.LENGTH_LONG).show();
                    Utility.setSharedPreference(context, Constants.KEY_CURRENT_LOC, s_address);


                    if (!from.equals("home")) {
                        Intent intent = getIntent();
                        intent.putExtra("currentTemp", currentTemp);
                        intent.putExtra("tempType", tempType);
                        setResult(111, intent);
                        finish();
                    } else {
                        Intent intent = new Intent(context, HomePage.class);
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                        finish();
                    }

                } else if (model != null && model.getError().equals(true)) {
                    Utility.stopProgressDialog(context);
                    Toast.makeText(context, model.getMessage(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<ModelResponse> call, Throwable t) {
                Utility.stopProgressDialog(context);
                Toast.makeText(context, Constants.ERROR_MSG, Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }
}

File: app/src/main/java/libertypassage/com/corporate/view/UpdateAddress.kt
package libertypassage.com.corporate.view

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.location.Address
import android.location.Geocoder
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.tasks.OnCompleteListener
import kotlinx.android.synthetic.main.activity_header.*
import kotlinx.android.synthetic.main.update_address.*
import libertypassage.com.corporate.R
import libertypassage.com.corporate.model.ModelResponse
import libertypassage.com.corporate.others.QRCodeScannerActivity
import libertypassage.com.corporate.retofit.ApiInterface
import libertypassage.com.corporate.retofit.ClientInstance
import libertypassage.com.corporate.utilities.Constants
import libertypassage.com.corporate.utilities.DialogProgress
import libertypassage.com.corporate.utilities.Utility
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.util.*
import java.util.regex.Matcher
import java.util.regex.Pattern


class UpdateAddress : AppCompatActivity(), View.OnClickListener {
    lateinit var context: Context
    var dialogProgress: DialogProgress? = null
    var mFusedLocationClient: FusedLocationProviderClient? = null
    private val urlRegex = "^((https?|ftp)://|(www|ftp)\\.)?[a-z0-9-]+(\\.[a-z0-9-]+)+([/?].*)?$"
    private var token = ""
    private var address = ""
    private var tempType = ""
    private var currentTemp = ""
    private var latitudes = 0.0
    private var longitudes = 0.0
    private var altitudes = 0.0
    private var from = ""
    private var venueId = ""
    private var nircId = ""



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.update_address)
        setupParent(window.decorView.findViewById(android.R.id.content))
        context = this@UpdateAddress

        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        dialogProgress = DialogProgress(context)
        token = Utility.getSharedPreferences(context, Constants.KEY_BEARER_TOKEN)!!
        tempType = Utility.getSharedPreferences(context, Constants.KEY_TEMP_TYPE)!!
        currentTemp = Utility.getSharedPreferences(context, Constants.KEY_MY_TEMP)!!
        nircId = Utility.getSharedPreferences(context, Constants.KEY_NRIC_ID)!!

        from = intent.getStringExtra("from")!!
        et_address.setText(Utility.getSharedPreferences(context, Constants.KEY_CURRENT_LOC))

        iv_back.setOnClickListener(this)
        iv_qrcode_scanerr.setOnClickListener(this)
        tv_currentLocation.setOnClickListener(this)
        iv_edit.setOnClickListener(this)
        tv_next.setOnClickListener(this)
    }


    override fun onClick(v: View?) {
        when (v!!.id) {
            R.id.iv_back -> {
                finish()
            }

            R.id.iv_edit -> {
                et_address.isEnabled = true
                et_address.setSelection(et_address.length())
            }

            R.id.iv_qrcode_scanerr -> {
                val intent = Intent(context, QRCodeScannerActivity::class.java)
                startActivityForResult(intent, 170)
            }

            R.id.tv_currentLocation -> {
                getLastLocation()
//                dialogUpdate()
            }

            R.id.tv_next -> {
                address = et_address.text.toString().trim()

                val pattern: Pattern = Pattern.compile(urlRegex)
                val matcher: Matcher = pattern.matcher(address)

                if (address == "") {
                    Toast.makeText(context, "Require address", Toast.LENGTH_LONG).show()
                } else if (Utility.isConnectingToInternet(context)) {
                    if (matcher.find() && address.length>50) {
                        venueId = address.substring(address.lastIndexOf("/") + 1)
                        val intent = Intent(context, ConfirmNricId::class.java)
                        intent.putExtra("from", from)
                        intent.putExtra("address", address)
                        intent.putExtra("venueId", venueId)
                        startActivity(intent)
                        finish()
                    } else {
                        addAddressAndTemp()
                    }
                } else {
                    Toast.makeText(context, "Please connect the internet", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

//    private fun dialogUpdate() {
//        val dialog = Dialog(context)
//        dialog.window!!.setLayout(
//            ViewGroup.LayoutParams.MATCH_PARENT,
//            ViewGroup.LayoutParams.WRAP_CONTENT
//        )
//        dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
//        dialog.setCancelable(true)
//        dialog.setContentView(R.layout.dialog_get_location)
//
//        val ivCancel = dialog.findViewById(R.id.iv_cancel) as ImageView
//        val tvNotAllow = dialog.findViewById(R.id.tv_notAllow) as TextView
//        val tvOk = dialog.findViewById(R.id.tv_ok) as TextView
//
//        ivCancel.setOnClickListener(View.OnClickListener {
//            dialog.dismiss()
//        })
//        tvNotAllow.setOnClickListener(View.OnClickListener {
//            dialog.dismiss()
//        })
//        tvOk.setOnClickListener(View.OnClickListener {
//            getLastLocation()
//            dialog.dismiss()
//        })
//        dialog.show()
//    }


    @SuppressLint("MissingPermission")
    private fun getLastLocation() {
        mFusedLocationClient?.lastLocation?.addOnCompleteListener(
            OnCompleteListener<Location?> { task ->
                val location = task.result
                if (location != null) {
                    latitudes = location.latitude
                    longitudes = location.longitude
                    altitudes = location.altitude
                    getAddress()
                }
            }
        )
    }

    private fun getAddress() {
        val addresses: List<Address>
        val geocoder: Geocoder = Geocoder(context, Locale.getDefault())
        try {
            addresses = geocoder.getFromLocation(
                latitudes,
                longitudes,
                1
            )

            val address11 = addresses[0].getAddressLine(0)
            address = address11
            et_address.setText(address)
            et_address.setSelection(et_address.length())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun addAddressAndTemp() {
        dialogProgress!!.show()
        val apiInterface: ApiInterface = ClientInstance.retrofitInstance!!.create(ApiInterface::class.java)
        val call: Call<ModelResponse> = apiInterface.addAddressTemp(
            Constants.KEY_HEADER+token,
            Constants.KEY_BOT, address, tempType, currentTemp, venueId, nircId, "checkin"
        )

        call.enqueue(object : Callback<ModelResponse?> {
            override fun onResponse(
                call: Call<ModelResponse?>,
                response: Response<ModelResponse?>
            ) {
                dialogProgress!!.dismiss()
                val responses: ModelResponse? = response.body()

                if (responses != null && responses.error!!.equals(false)) {
                    Toast.makeText(context, responses.message, Toast.LENGTH_LONG).show()
                    Utility.setSharedPreference(context, Constants.KEY_CURRENT_LOC, address)

                    if (from != "home") {
                        val intent = intent
                        intent.putExtra("currentTemp", currentTemp)
                        intent.putExtra("tempType", tempType)
                        setResult(111, intent)
                        finish()
                    } else {
                        finish()
                    }

                } else if (responses != null && responses.error!!.equals(true)) {
                    dialogProgress!!.dismiss()
                    Toast.makeText(context, responses.message, Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<ModelResponse?>, t: Throwable) {
                dialogProgress!!.dismiss()
                Log.e("model", "onFailure    " + t.message)
            }
        })
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 170) {
            if (resultCode == 160) {
                val contents = data?.getStringExtra("ResultText")
                if (contents != null) {
                    address = contents
                    et_address.setText(address)
                    et_address.setSelection(et_address.length())
                    Log.e("contents", contents)
                }
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupParent(view: View) {
        if (view !is EditText) {
            view.setOnTouchListener { v, event ->
                if (currentFocus != null) {
                    val inputMethodManager =
                        getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                    inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
                }
                false
            }
        }
        //If a layout container, iterate over children
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                val innerView = view.getChildAt(i)
                setupParent(innerView)
            }
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }
}
